// goey-toast/dist/index.js
import { useState, useEffect, Component, useRef, useCallback, useLayoutEffect } from 'react';
import { toast, Toaster } from 'sonner';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { animate, AnimatePresence, motion } from 'framer-motion';

// src/components/GoeyToaster.tsx

// src/context.ts
var _position = "bottom-right";
var _spring = true;
var _bounce = void 0;
function setGoeyPosition(position) {
  _position = position;
}
function getGoeyPosition() {
  return _position;
}
function setGoeySpring(spring) {
  _spring = spring;
}
function getGoeySpring() {
  return _spring;
}
function setGoeyBounce(bounce) {
  _bounce = bounce;
}
function getGoeyBounce() {
  return _bounce;
}
function GoeyToaster({
  position = "bottom-right",
  duration,
  gap = 14,
  offset = "24px",
  theme = "light",
  toastOptions,
  expand,
  closeButton,
  richColors,
  visibleToasts,
  dir,
  spring = true,
  bounce
}) {
  useEffect(() => {
    setGoeyPosition(position);
  }, [position]);
  useEffect(() => {
    setGoeySpring(spring);
  }, [spring]);
  useEffect(() => {
    setGoeyBounce(bounce);
  }, [bounce]);
  useEffect(() => {
    if (process.env.NODE_ENV !== "development") return;
    const el = document.createElement("div");
    el.setAttribute("data-goey-toast-css", "");
    el.style.position = "absolute";
    el.style.width = "0";
    el.style.height = "0";
    el.style.overflow = "hidden";
    el.style.pointerEvents = "none";
    document.body.appendChild(el);
    const value = getComputedStyle(el).getPropertyValue("--goey-toast");
    document.body.removeChild(el);
    if (!value) {
      console.warn(
        '[goey-toast] Styles not found. Make sure to import the CSS:\n\n  import "goey-toast/styles.css";\n'
      );
    }
  }, []);
  return /* @__PURE__ */ jsx(
    Toaster,
    {
      position,
      duration,
      gap,
      offset,
      theme,
      toastOptions: { unstyled: true, ...toastOptions },
      expand,
      closeButton,
      richColors,
      visibleToasts,
      dir
    }
  );
}
var DefaultIcon = ({ size = 20, className }) => /* @__PURE__ */ jsxs(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: size,
    height: size,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    className,
    children: [
      /* @__PURE__ */ jsx("path", { d: "M18 8A6 6 0 0 0 6 8c0 7-3 9-3 9h18s-3-2-3-9" }),
      /* @__PURE__ */ jsx("path", { d: "M13.73 21a2 2 0 0 1-3.46 0" })
    ]
  }
);
var SuccessIcon = ({ size = 20, className }) => /* @__PURE__ */ jsxs(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: size,
    height: size,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    className,
    children: [
      /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "10" }),
      /* @__PURE__ */ jsx("path", { d: "M9 12l2 2 4-4" })
    ]
  }
);
var ErrorIcon = ({ size = 20, className }) => /* @__PURE__ */ jsxs(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: size,
    height: size,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    className,
    children: [
      /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "10" }),
      /* @__PURE__ */ jsx("path", { d: "M15 9l-6 6" }),
      /* @__PURE__ */ jsx("path", { d: "M9 9l6 6" })
    ]
  }
);
var WarningIcon = ({ size = 20, className }) => /* @__PURE__ */ jsxs(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: size,
    height: size,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    className,
    children: [
      /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "10" }),
      /* @__PURE__ */ jsx("line", { x1: "12", y1: "8", x2: "12", y2: "12" }),
      /* @__PURE__ */ jsx("line", { x1: "12", y1: "16", x2: "12.01", y2: "16" })
    ]
  }
);
var InfoIcon = ({ size = 20, className }) => /* @__PURE__ */ jsxs(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: size,
    height: size,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    className,
    children: [
      /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "10" }),
      /* @__PURE__ */ jsx("line", { x1: "12", y1: "16", x2: "12", y2: "12" }),
      /* @__PURE__ */ jsx("line", { x1: "12", y1: "8", x2: "12.01", y2: "8" })
    ]
  }
);
var SpinnerIcon = ({ size = 20, className }) => /* @__PURE__ */ jsx(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: size,
    height: size,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    className,
    children: /* @__PURE__ */ jsx("path", { d: "M21 12a9 9 0 1 1-6.219-8.56" })
  }
);
var QUERY = "(prefers-reduced-motion: reduce)";
function getInitialState() {
  if (typeof window === "undefined" || typeof window.matchMedia !== "function") {
    return false;
  }
  return window.matchMedia(QUERY).matches;
}
function usePrefersReducedMotion() {
  const [prefersReducedMotion, setPrefersReducedMotion] = useState(getInitialState);
  useEffect(() => {
    if (typeof window === "undefined" || typeof window.matchMedia !== "function") {
      return;
    }
    const mql = window.matchMedia(QUERY);
    const handler = (event) => {
      setPrefersReducedMotion(event.matches);
    };
    mql.addEventListener("change", handler);
    return () => mql.removeEventListener("change", handler);
  }, []);
  return prefersReducedMotion;
}

// src/components/goey-styles.ts
var styles = {
  spinnerSpin: "goey-spinnerSpin",
  wrapper: "goey-wrapper",
  blobSvg: "goey-blobSvg",
  content: "goey-content",
  contentCompact: "goey-contentCompact",
  contentExpanded: "goey-contentExpanded",
  header: "goey-header",
  iconWrapper: "goey-iconWrapper",
  title: "goey-title",
  titleDefault: "goey-titleDefault",
  titleSuccess: "goey-titleSuccess",
  titleError: "goey-titleError",
  titleWarning: "goey-titleWarning",
  titleInfo: "goey-titleInfo",
  titleLoading: "goey-titleLoading",
  description: "goey-description",
  actionWrapper: "goey-actionWrapper",
  actionButton: "goey-actionButton",
  actionDefault: "goey-actionDefault",
  actionSuccess: "goey-actionSuccess",
  actionError: "goey-actionError",
  actionWarning: "goey-actionWarning",
  actionInfo: "goey-actionInfo"
};
var phaseIconMap = {
  default: DefaultIcon,
  success: SuccessIcon,
  error: ErrorIcon,
  warning: WarningIcon,
  info: InfoIcon
};
var titleColorMap = {
  loading: styles.titleLoading,
  default: styles.titleDefault,
  success: styles.titleSuccess,
  error: styles.titleError,
  warning: styles.titleWarning,
  info: styles.titleInfo
};
var actionColorMap = {
  loading: styles.actionInfo,
  default: styles.actionDefault,
  success: styles.actionSuccess,
  error: styles.actionError,
  warning: styles.actionWarning,
  info: styles.actionInfo
};
var PH = 34;
var DEFAULT_DISPLAY_DURATION = 4e3;
var DEFAULT_EXPAND_DUR = 0.6;
var DEFAULT_COLLAPSE_DUR = 0.9;
function squishSpring(durationSec, defaultDur, bounce = 0.4) {
  const scale = durationSec / defaultDur;
  const stiffness = 200 + bounce * 437.5;
  const damping = 24 - bounce * 20;
  const mass = 0.7 * scale;
  return { type: "spring", stiffness, damping, mass };
}
var observerRegistry = /* @__PURE__ */ new Map();
function registerSonnerObserver(ol, callback) {
  let entry = observerRegistry.get(ol);
  if (!entry) {
    const callbacks = /* @__PURE__ */ new Set();
    let applying = false;
    const observer = new MutationObserver(() => {
      if (applying) return;
      applying = true;
      requestAnimationFrame(() => {
        callbacks.forEach((cb) => cb());
        requestAnimationFrame(() => {
          applying = false;
        });
      });
    });
    observer.observe(ol, {
      attributes: true,
      attributeFilter: ["style"],
      subtree: true,
      childList: true
    });
    entry = { observer, callbacks };
    observerRegistry.set(ol, entry);
  }
  entry.callbacks.add(callback);
  return () => {
    entry.callbacks.delete(callback);
    if (entry.callbacks.size === 0) {
      entry.observer.disconnect();
      observerRegistry.delete(ol);
    }
  };
}
function syncSonnerHeights(wrapperEl) {
  if (!wrapperEl) return;
  const li = wrapperEl.closest("[data-sonner-toast]");
  if (!li?.parentElement) return;
  const ol = li.parentElement;
  const toasts = Array.from(
    ol.querySelectorAll(":scope > [data-sonner-toast]")
  );
  for (const t of toasts) {
    const content = t.firstElementChild;
    const height = content ? content.getBoundingClientRect().height : 0;
    if (height > 0) {
      t.style.setProperty("--initial-height", `${height}px`);
    }
  }
}
function morphPath(pw, bw, th, t) {
  const pr = PH / 2;
  const pillW = Math.min(pw, bw);
  const bodyH = PH + (th - PH) * t;
  if (t <= 0 || bodyH - PH < 8) {
    return [
      `M 0,${pr}`,
      `A ${pr},${pr} 0 0 1 ${pr},0`,
      `H ${pillW - pr}`,
      `A ${pr},${pr} 0 0 1 ${pillW},${pr}`,
      `A ${pr},${pr} 0 0 1 ${pillW - pr},${PH}`,
      `H ${pr}`,
      `A ${pr},${pr} 0 0 1 0,${pr}`,
      `Z`
    ].join(" ");
  }
  const curve = 14 * t;
  const cr = Math.min(16, (bodyH - PH) * 0.45);
  const bodyW = pillW + (bw - pillW) * t;
  const bodyTop = PH - curve;
  const qEndX = Math.min(pillW + curve, bodyW - cr);
  return [
    `M 0,${pr}`,
    `A ${pr},${pr} 0 0 1 ${pr},0`,
    `H ${pillW - pr}`,
    `A ${pr},${pr} 0 0 1 ${pillW},${pr}`,
    `L ${pillW},${bodyTop}`,
    `Q ${pillW},${bodyTop + curve} ${qEndX},${bodyTop + curve}`,
    `H ${bodyW - cr}`,
    `A ${cr},${cr} 0 0 1 ${bodyW},${bodyTop + curve + cr}`,
    `L ${bodyW},${bodyH - cr}`,
    `A ${cr},${cr} 0 0 1 ${bodyW - cr},${bodyH}`,
    `H ${cr}`,
    `A ${cr},${cr} 0 0 1 0,${bodyH - cr}`,
    `Z`
  ].join(" ");
}
function morphPathCenter(pw, bw, th, t) {
  const pr = PH / 2;
  const pillW = Math.min(pw, bw);
  const pillOffset = (bw - pillW) / 2;
  if (t <= 0 || PH + (th - PH) * t - PH < 8) {
    return [
      `M ${pillOffset},${pr}`,
      `A ${pr},${pr} 0 0 1 ${pillOffset + pr},0`,
      `H ${pillOffset + pillW - pr}`,
      `A ${pr},${pr} 0 0 1 ${pillOffset + pillW},${pr}`,
      `A ${pr},${pr} 0 0 1 ${pillOffset + pillW - pr},${PH}`,
      `H ${pillOffset + pr}`,
      `A ${pr},${pr} 0 0 1 ${pillOffset},${pr}`,
      `Z`
    ].join(" ");
  }
  const bodyH = PH + (th - PH) * t;
  const curve = 14 * t;
  const cr = Math.min(16, (bodyH - PH) * 0.45);
  const bodyTop = PH - curve;
  const bodyCenter = bw / 2;
  const halfBodyW = pillW / 2 + (bw - pillW) / 2 * t;
  const bodyLeft = bodyCenter - halfBodyW;
  const bodyRight = bodyCenter + halfBodyW;
  const qLeftX = Math.max(bodyLeft + cr, pillOffset - curve);
  const qRightX = Math.min(bodyRight - cr, pillOffset + pillW + curve);
  return [
    // Start at left side of pill
    `M ${pillOffset},${pr}`,
    // Pill top-left arc
    `A ${pr},${pr} 0 0 1 ${pillOffset + pr},0`,
    // Top edge of pill
    `H ${pillOffset + pillW - pr}`,
    // Pill top-right arc
    `A ${pr},${pr} 0 0 1 ${pillOffset + pillW},${pr}`,
    // Right side down to body junction
    `L ${pillOffset + pillW},${bodyTop}`,
    // Right organic curve from pill to body
    `Q ${pillOffset + pillW},${bodyTop + curve} ${qRightX},${bodyTop + curve}`,
    // Right side of body
    `H ${bodyRight - cr}`,
    // Body top-right corner
    `A ${cr},${cr} 0 0 1 ${bodyRight},${bodyTop + curve + cr}`,
    // Right edge down
    `L ${bodyRight},${bodyH - cr}`,
    // Body bottom-right corner
    `A ${cr},${cr} 0 0 1 ${bodyRight - cr},${bodyH}`,
    // Bottom edge
    `H ${bodyLeft + cr}`,
    // Body bottom-left corner
    `A ${cr},${cr} 0 0 1 ${bodyLeft},${bodyH - cr}`,
    // Left edge up
    `L ${bodyLeft},${bodyTop + curve + cr}`,
    // Body top-left corner
    `A ${cr},${cr} 0 0 1 ${bodyLeft + cr},${bodyTop + curve}`,
    // Left side of body
    `H ${qLeftX}`,
    // Left organic curve from body to pill
    `Q ${pillOffset},${bodyTop + curve} ${pillOffset},${bodyTop}`,
    // Close back to start
    `Z`
  ].join(" ");
}
var SMOOTH_EASE = [0.4, 0, 0.2, 1];
var GoeyToast = ({
  title,
  description,
  action,
  icon,
  phase,
  classNames,
  fillColor = "#ffffff",
  borderColor,
  borderWidth,
  timing,
  spring: springProp,
  bounce: bounceProp,
  toastId
}) => {
  const position = getGoeyPosition();
  const isRight = position?.includes("right") ?? false;
  const isCenter = position?.includes("center") ?? false;
  const prefersReducedMotion = usePrefersReducedMotion();
  const useSpring = springProp ?? getGoeySpring();
  const bounceVal = bounceProp ?? getGoeyBounce() ?? 0.4;
  const [actionSuccess, setActionSuccess] = useState(null);
  const [dismissing, setDismissing] = useState(false);
  const [hovered, setHovered] = useState(false);
  const hoveredRef = useRef(false);
  const collapsingRef = useRef(false);
  const preDismissRef = useRef(false);
  const collapseEndTime = useRef(0);
  const expandedDimsRef = useRef({ pw: 0, bw: 0, th: 0 });
  const dismissTimerRef = useRef(null);
  const effectiveTitle = actionSuccess ?? title;
  const effectivePhase = actionSuccess ? "success" : phase;
  const effectiveDescription = actionSuccess ? void 0 : description;
  const effectiveAction = actionSuccess ? void 0 : action;
  const isLoading = effectivePhase === "loading";
  const hasDescription = Boolean(effectiveDescription);
  const hasAction = Boolean(effectiveAction);
  const isExpanded = (hasDescription || hasAction) && !dismissing;
  const [showBody, setShowBody] = useState(false);
  const wrapperRef = useRef(null);
  const pathRef = useRef(null);
  const headerRef = useRef(null);
  const contentRef = useRef(null);
  const morphCtrl = useRef(null);
  const pillResizeCtrl = useRef(null);
  const headerSquishCtrl = useRef(null);
  const morphTRef = useRef(0);
  const aDims = useRef({ pw: 0, bw: 0, th: 0 });
  const dimsRef = useRef({ pw: 0, bw: 0, th: 0 });
  const [dims, setDims] = useState({ pw: 0, bw: 0, th: 0 });
  useEffect(() => {
    dimsRef.current = dims;
  }, [dims]);
  const flush = useCallback(() => {
    const { pw: p, bw: b, th: h } = aDims.current;
    if (p <= 0 || b <= 0 || h <= 0) return;
    const t = Math.max(0, Math.min(1, morphTRef.current));
    const pos = getGoeyPosition();
    const rightSide = pos?.includes("right") ?? false;
    const centerPos = pos?.includes("center") ?? false;
    if (centerPos) {
      const centerBw = Math.max(dimsRef.current.bw, expandedDimsRef.current.bw, p);
      pathRef.current?.setAttribute("d", morphPathCenter(p, centerBw, h, t));
    } else {
      pathRef.current?.setAttribute("d", morphPath(p, b, h, t));
    }
    if (t >= 1) {
      if (wrapperRef.current) {
        wrapperRef.current.style.width = "";
      }
      if (contentRef.current) {
        contentRef.current.style.width = "";
        contentRef.current.style.overflow = "";
        contentRef.current.style.maxHeight = "";
        contentRef.current.style.clipPath = "";
      }
    } else if (t > 0) {
      const targetBw = dimsRef.current.bw;
      const targetTh = dimsRef.current.th;
      const pillW = Math.min(p, b);
      const currentW = pillW + (b - pillW) * t;
      const currentH = PH + (targetTh - PH) * t;
      const centerFullW = centerPos ? Math.max(dimsRef.current.bw, expandedDimsRef.current.bw, p) : 0;
      if (wrapperRef.current) {
        wrapperRef.current.style.width = (centerPos ? centerFullW : currentW) + "px";
      }
      if (contentRef.current) {
        contentRef.current.style.width = (centerPos ? centerFullW : targetBw) + "px";
        contentRef.current.style.overflow = "hidden";
        contentRef.current.style.maxHeight = currentH + "px";
        if (centerPos) {
          const clip = (centerFullW - currentW) / 2;
          contentRef.current.style.clipPath = `inset(0 ${clip}px 0 ${clip}px)`;
        } else {
          const clip = targetBw - currentW;
          contentRef.current.style.clipPath = rightSide ? `inset(0 0 0 ${clip}px)` : `inset(0 ${clip}px 0 0)`;
        }
      }
    } else {
      const pillW = Math.min(p, b);
      if (wrapperRef.current) {
        const centerBw = centerPos ? Math.max(dimsRef.current.bw, expandedDimsRef.current.bw, p) : pillW;
        wrapperRef.current.style.width = centerBw + "px";
      }
      if (contentRef.current) {
        if (centerPos) {
          const centerBwVal = Math.max(dimsRef.current.bw, expandedDimsRef.current.bw, p);
          contentRef.current.style.width = centerBwVal + "px";
          const clip = (centerBwVal - pillW) / 2;
          contentRef.current.style.clipPath = `inset(0 ${clip}px 0 ${clip}px)`;
        } else {
          contentRef.current.style.width = "";
          contentRef.current.style.clipPath = "";
        }
        contentRef.current.style.overflow = "hidden";
        contentRef.current.style.maxHeight = PH + "px";
      }
    }
  }, []);
  const measure = useCallback(() => {
    if (!headerRef.current || !contentRef.current) return;
    const wr = wrapperRef.current;
    const savedW = wr?.style.width ?? "";
    const savedOv = contentRef.current.style.overflow;
    const savedMH = contentRef.current.style.maxHeight;
    const savedCW = contentRef.current.style.width;
    if (wr) {
      wr.style.width = "";
    }
    contentRef.current.style.overflow = "";
    contentRef.current.style.maxHeight = "";
    contentRef.current.style.width = "";
    const cs = getComputedStyle(contentRef.current);
    const paddingX = parseFloat(cs.paddingLeft) + parseFloat(cs.paddingRight);
    const pw2 = headerRef.current.offsetWidth + paddingX;
    const bw2 = contentRef.current.offsetWidth;
    const th2 = contentRef.current.offsetHeight;
    if (wr) {
      wr.style.width = savedW;
    }
    contentRef.current.style.overflow = savedOv;
    contentRef.current.style.maxHeight = savedMH;
    contentRef.current.style.width = savedCW;
    dimsRef.current = { pw: pw2, bw: bw2, th: th2 };
    setDims({ pw: pw2, bw: bw2, th: th2 });
  }, []);
  useLayoutEffect(() => {
    measure();
    const t = setTimeout(measure, 100);
    return () => clearTimeout(t);
  }, [effectiveTitle, effectivePhase, isExpanded, showBody, effectiveDescription, effectiveAction, measure]);
  useEffect(() => {
    if (!contentRef.current) return;
    const ro = new ResizeObserver(measure);
    ro.observe(contentRef.current);
    return () => ro.disconnect();
  }, [measure]);
  const { pw, bw, th } = dims;
  const hasDims = pw > 0 && bw > 0 && th > 0;
  const blobSquishCtrl = useRef(null);
  const expandDur = DEFAULT_EXPAND_DUR;
  const collapseDur = DEFAULT_COLLAPSE_DUR;
  const lastSquishTime = useRef(0);
  const triggerLandingSquish = useCallback((phase2 = "mount") => {
    if (!wrapperRef.current || prefersReducedMotion) return;
    if (!useSpring) return;
    const now = Date.now();
    if (now - lastSquishTime.current < 300) return;
    lastSquishTime.current = now;
    blobSquishCtrl.current?.stop();
    const el = wrapperRef.current;
    const springConfig = phase2 === "collapse" ? squishSpring(collapseDur, DEFAULT_COLLAPSE_DUR, bounceVal) : squishSpring(expandDur, DEFAULT_EXPAND_DUR, bounceVal);
    const bScale = bounceVal / 0.4;
    const compressY = (phase2 === "collapse" ? 0.07 : 0.12) * bScale;
    const expandX = (phase2 === "collapse" ? 0.035 : 0.06) * bScale;
    blobSquishCtrl.current = animate(0, 1, {
      ...springConfig,
      onUpdate: (v) => {
        const intensity = Math.sin(v * Math.PI);
        const sy = 1 - compressY * intensity;
        const sx = 1 + expandX * intensity;
        const mirror = el.style.transform?.includes("scaleX(-1)") ? "scaleX(-1) " : "";
        el.style.transformOrigin = "center top";
        el.style.transform = mirror + `scaleX(${sx}) scaleY(${sy})`;
      },
      onComplete: () => {
        const right = el.style.transform?.includes("scaleX(-1)");
        el.style.transform = right ? "scaleX(-1)" : "";
        el.style.transformOrigin = "";
      }
    });
  }, [prefersReducedMotion, expandDur, collapseDur, useSpring, bounceVal]);
  useLayoutEffect(() => {
    if (!hasDims || collapsingRef.current) return;
    const prev = { ...aDims.current };
    const target = { pw, bw, th };
    if (prev.bw <= 0) {
      aDims.current = target;
      flush();
      return;
    }
    if (morphTRef.current > 0 && morphTRef.current < 1) {
      aDims.current = target;
      flush();
      return;
    }
    if (showBody) {
      aDims.current = target;
      flush();
      return;
    }
    if (prev.bw === target.bw && prev.pw === target.pw && prev.th === target.th) return;
    if (prefersReducedMotion) {
      aDims.current = target;
      flush();
      return;
    }
    pillResizeCtrl.current?.stop();
    if (Date.now() - collapseEndTime.current > 500 && !isExpanded) {
      triggerLandingSquish("expand");
    }
    const pillResizeTransition = useSpring ? { type: "spring", duration: 0.5, bounce: bounceVal * 0.875 } : { duration: 0.4, ease: SMOOTH_EASE };
    pillResizeCtrl.current = animate(0, 1, {
      ...pillResizeTransition,
      onUpdate: (t) => {
        aDims.current = {
          pw: prev.pw + (target.pw - prev.pw) * t,
          bw: prev.bw + (target.bw - prev.bw) * t,
          th: prev.th + (target.th - prev.th) * t
        };
        flush();
      }
    });
  }, [pw, bw, th, hasDims, showBody, flush, prefersReducedMotion, triggerLandingSquish, useSpring]);
  const squishDelayMs = 45;
  const mountSquished = useRef(false);
  useEffect(() => {
    if (hasDims && !mountSquished.current && !isExpanded) {
      mountSquished.current = true;
      const t = setTimeout(triggerLandingSquish, squishDelayMs);
      return () => clearTimeout(t);
    }
  }, [hasDims, squishDelayMs, triggerLandingSquish]);
  const prevShowBody = useRef(false);
  useLayoutEffect(() => {
    if (!prevShowBody.current && showBody && !hoveredRef.current) {
      const t = setTimeout(() => triggerLandingSquish("expand"), 80);
      return () => clearTimeout(t);
    }
    prevShowBody.current = showBody;
  }, [showBody, triggerLandingSquish]);
  const shakeCtrl = useRef(null);
  const prevPhase = useRef(phase);
  useEffect(() => {
    if (phase === "error" && prevPhase.current !== "error" && !dismissing && wrapperRef.current && !prefersReducedMotion) {
      shakeCtrl.current?.stop();
      const el = wrapperRef.current;
      const mirror = el.style.transform?.includes("scaleX(-1)") ? "scaleX(-1) " : "";
      shakeCtrl.current = animate(0, 1, {
        duration: 0.4,
        ease: "easeOut",
        onUpdate: (v) => {
          const decay = 1 - v;
          const shake = Math.sin(v * Math.PI * 6) * decay * 3;
          el.style.transform = mirror + `translateX(${shake}px)`;
        },
        onComplete: () => {
          el.style.transform = mirror.trim() || "";
        }
      });
    }
    prevPhase.current = phase;
    return () => {
      shakeCtrl.current?.stop();
    };
  }, [phase, dismissing, prefersReducedMotion]);
  useEffect(() => {
    if (isExpanded) {
      const delay = prefersReducedMotion ? 0 : 330;
      const t1 = setTimeout(() => setShowBody(true), delay);
      return () => clearTimeout(t1);
    }
    morphCtrl.current?.stop();
    pillResizeCtrl.current?.stop();
    if (morphTRef.current > 0) {
      const csPad = contentRef.current ? getComputedStyle(contentRef.current) : null;
      const padX = csPad ? parseFloat(csPad.paddingLeft) + parseFloat(csPad.paddingRight) : 20;
      const targetPw = headerRef.current ? headerRef.current.offsetWidth + padX : aDims.current.pw;
      const targetDims = { pw: targetPw, bw: targetPw, th: PH };
      if (prefersReducedMotion) {
        morphTRef.current = 0;
        collapsingRef.current = false;
        preDismissRef.current = false;
        setShowBody(false);
        aDims.current = { ...targetDims };
        flush();
        return;
      }
      const savedDims = expandedDimsRef.current.bw > 0 ? { ...expandedDimsRef.current } : { ...aDims.current };
      const isPreDismiss = preDismissRef.current;
      const collapseDur2 = 0.9;
      const collapseTransition = isPreDismiss || !useSpring ? { duration: collapseDur2, ease: SMOOTH_EASE } : { type: "spring", duration: collapseDur2, bounce: bounceVal * 0.875 };
      triggerLandingSquish("collapse");
      morphCtrl.current = animate(morphTRef.current, 0, {
        ...collapseTransition,
        onUpdate: (t) => {
          morphTRef.current = t;
          aDims.current = {
            pw: targetDims.pw + (savedDims.pw - targetDims.pw) * t,
            bw: targetDims.bw + (savedDims.bw - targetDims.bw) * t,
            th: targetDims.th + (savedDims.th - targetDims.th) * t
          };
          flush();
        },
        onComplete: () => {
          morphTRef.current = 0;
          collapsingRef.current = false;
          preDismissRef.current = false;
          collapseEndTime.current = Date.now();
          aDims.current = { ...targetDims };
          flush();
          setShowBody(false);
        }
      });
      return () => {
        morphCtrl.current?.stop();
      };
    }
    setShowBody(false);
    morphTRef.current = 0;
    flush();
  }, [isExpanded, flush, prefersReducedMotion, useSpring, triggerLandingSquish]);
  const remainingRef = useRef(null);
  const timerStartRef = useRef(0);
  useEffect(() => {
    if (!showBody || actionSuccess || dismissing) return;
    const expandDelayMs = prefersReducedMotion ? 0 : 330;
    const collapseMs = prefersReducedMotion ? 10 : 0.9 * 1e3;
    const displayMs = timing?.displayDuration ?? DEFAULT_DISPLAY_DURATION;
    const fullDelay = displayMs - expandDelayMs - collapseMs;
    if (fullDelay <= 0) return;
    if (hoveredRef.current) return;
    const delay = remainingRef.current ?? fullDelay;
    timerStartRef.current = Date.now();
    const timer = setTimeout(() => {
      remainingRef.current = null;
      expandedDimsRef.current = { ...aDims.current };
      collapsingRef.current = true;
      preDismissRef.current = true;
      setDismissing(true);
    }, delay);
    dismissTimerRef.current = timer;
    return () => {
      clearTimeout(timer);
      const elapsed = Date.now() - timerStartRef.current;
      const remaining = delay - elapsed;
      if (remaining > 0 && hoveredRef.current) {
        remainingRef.current = remaining;
      }
    };
  }, [showBody, actionSuccess, dismissing, prefersReducedMotion, hovered]);
  const canExpand = hasDescription || hasAction;
  const reExpandingRef = useRef(false);
  useEffect(() => {
    if (!hovered || !canExpand || !dismissing) return;
    morphCtrl.current?.stop();
    collapsingRef.current = false;
    preDismissRef.current = false;
    remainingRef.current = null;
    reExpandingRef.current = true;
    setDismissing(false);
    setShowBody(true);
    const currentT = morphTRef.current;
    const startDims = { ...aDims.current };
    const morphExpandTransition = useSpring ? { type: "spring", duration: 0.9, bounce: bounceVal } : { duration: 0.6, ease: SMOOTH_EASE };
    requestAnimationFrame(() => {
      morphCtrl.current = animate(currentT, 1, {
        ...morphExpandTransition,
        onUpdate: (t) => {
          morphTRef.current = t;
          const target = dimsRef.current;
          aDims.current = {
            pw: startDims.pw + (target.pw - startDims.pw) * t,
            bw: startDims.bw + (target.bw - startDims.bw) * t,
            th: startDims.th + (target.th - startDims.th) * t
          };
          flush();
        },
        onComplete: () => {
          morphTRef.current = 1;
          aDims.current = { ...dimsRef.current };
          reExpandingRef.current = false;
          flush();
          syncSonnerHeights(wrapperRef.current);
        }
      });
    });
    return () => {
      morphCtrl.current?.stop();
    };
  }, [hovered, dismissing, canExpand]);
  useEffect(() => {
    if (!toastId || !dismissing || showBody || hovered) return;
    const t = setTimeout(() => {
      if (!hoveredRef.current) toast.dismiss(toastId);
    }, 800);
    return () => clearTimeout(t);
  }, [dismissing, showBody, hovered, toastId]);
  useEffect(() => {
    if (!toastId || !actionSuccess || showBody) return;
    const t = setTimeout(() => toast.dismiss(toastId), 1200);
    return () => clearTimeout(t);
  }, [toastId, actionSuccess, showBody]);
  useEffect(() => {
    if (reExpandingRef.current) return;
    if (!showBody) {
      morphTRef.current = 0;
      morphCtrl.current?.stop();
      flush();
      return;
    }
    if (prefersReducedMotion) {
      pillResizeCtrl.current?.stop();
      morphCtrl.current?.stop();
      morphTRef.current = 1;
      aDims.current = { ...dimsRef.current };
      flush();
      syncSonnerHeights(wrapperRef.current);
      return;
    }
    const raf = requestAnimationFrame(() => {
      pillResizeCtrl.current?.stop();
      morphCtrl.current?.stop();
      const startDims = { ...aDims.current };
      const morphExpandTransition = useSpring ? { type: "spring", duration: 0.9, bounce: bounceVal } : { duration: 0.6, ease: SMOOTH_EASE };
      morphCtrl.current = animate(0, 1, {
        ...morphExpandTransition,
        onUpdate: (t) => {
          morphTRef.current = t;
          const target = dimsRef.current;
          aDims.current = {
            pw: startDims.pw + (target.pw - startDims.pw) * t,
            bw: startDims.bw + (target.bw - startDims.bw) * t,
            th: startDims.th + (target.th - startDims.th) * t
          };
          flush();
        },
        onComplete: () => {
          morphTRef.current = 1;
          aDims.current = { ...dimsRef.current };
          flush();
          syncSonnerHeights(wrapperRef.current);
        }
      });
    });
    return () => {
      cancelAnimationFrame(raf);
      morphCtrl.current?.stop();
    };
  }, [showBody, flush, prefersReducedMotion, useSpring]);
  const headerSquished = useRef(false);
  useEffect(() => {
    if (!headerRef.current || prefersReducedMotion) return;
    headerSquishCtrl.current?.stop();
    const el = headerRef.current;
    if (showBody && !dismissing && !actionSuccess) {
      if (!useSpring) return;
      headerSquished.current = true;
      headerSquishCtrl.current = animate(0, 1, {
        ...squishSpring(expandDur, DEFAULT_EXPAND_DUR, bounceVal),
        onUpdate: (v) => {
          const scale = 1 - 0.05 * v;
          const pushY = v * 1;
          el.style.transform = `scale(${scale}) translateY(${pushY}px)`;
        }
      });
    } else if (headerSquished.current) {
      headerSquished.current = false;
      const isSpringCollapse = !preDismissRef.current && useSpring;
      const transition = isSpringCollapse ? squishSpring(collapseDur, DEFAULT_COLLAPSE_DUR, bounceVal) : { duration: collapseDur * 0.5, ease: SMOOTH_EASE };
      headerSquishCtrl.current = animate(1, 0, {
        ...transition,
        onUpdate: (v) => {
          const scale = 1 - 0.05 * v;
          const pushY = v * 1;
          el.style.transform = `scale(${scale}) translateY(${pushY}px)`;
        },
        onComplete: () => {
          el.style.transform = "";
        }
      });
    }
    return () => {
      headerSquishCtrl.current?.stop();
    };
  }, [showBody, dismissing, actionSuccess, prefersReducedMotion, expandDur, collapseDur, useSpring]);
  useEffect(() => {
    const wrapper = wrapperRef.current;
    if (!wrapper) return;
    const ol = wrapper.closest("[data-sonner-toast]")?.parentElement;
    if (!ol) return;
    return registerSonnerObserver(ol, () => {
      syncSonnerHeights(wrapper);
    });
  }, []);
  const handleActionClick = useCallback(() => {
    if (!effectiveAction) return;
    if (effectiveAction.successLabel) {
      expandedDimsRef.current = { ...aDims.current };
      collapsingRef.current = true;
      setActionSuccess(effectiveAction.successLabel);
    }
    try {
      effectiveAction.onClick();
    } catch {
    }
  }, [effectiveAction]);
  const renderIcon = () => {
    if (!actionSuccess && icon) return icon;
    if (isLoading) return /* @__PURE__ */ jsx(SpinnerIcon, { size: 18, className: styles.spinnerSpin });
    const IconComponent = phaseIconMap[effectivePhase];
    return /* @__PURE__ */ jsx(IconComponent, { size: 18 });
  };
  const iconTransition = prefersReducedMotion ? { duration: 0.01 } : { duration: 0.2 };
  const iconEl = /* @__PURE__ */ jsx("div", { className: `${styles.iconWrapper}${classNames?.icon ? ` ${classNames.icon}` : ""}`, children: /* @__PURE__ */ jsx(AnimatePresence, { mode: "wait", children: /* @__PURE__ */ jsx(
    motion.div,
    {
      initial: prefersReducedMotion ? false : { opacity: 0, scale: 0.5 },
      animate: { opacity: 1, scale: 1 },
      exit: { opacity: 0, scale: 0.5 },
      transition: iconTransition,
      children: renderIcon()
    },
    isLoading ? "spinner" : effectivePhase
  ) }) });
  const titleEl = /* @__PURE__ */ jsx("span", { className: `${styles.title}${classNames?.title ? ` ${classNames.title}` : ""}`, children: effectiveTitle });
  const iconAndTitle = /* @__PURE__ */ jsxs(Fragment, { children: [
    iconEl,
    titleEl
  ] });
  return /* @__PURE__ */ jsxs("div", { ref: wrapperRef, className: `${styles.wrapper}${classNames?.wrapper ? ` ${classNames.wrapper}` : ""}`, style: isCenter ? { margin: "0 auto" } : isRight ? { marginLeft: "auto", transform: "scaleX(-1)" } : void 0, role: effectivePhase === "error" ? "alert" : "status", "aria-live": effectivePhase === "error" ? "assertive" : "polite", "aria-atomic": "true", onMouseEnter: () => {
    hoveredRef.current = true;
    setHovered(true);
  }, onMouseLeave: () => {
    hoveredRef.current = false;
    setHovered(false);
  }, "data-center": isCenter || void 0, children: [
    /* @__PURE__ */ jsx(
      "svg",
      {
        className: styles.blobSvg,
        "aria-hidden": true,
        children: /* @__PURE__ */ jsx(
          "path",
          {
            ref: pathRef,
            fill: fillColor,
            stroke: borderColor || "none",
            strokeWidth: borderColor ? borderWidth ?? 1.5 : 0
          }
        )
      }
    ),
    /* @__PURE__ */ jsxs(
      "div",
      {
        ref: contentRef,
        className: `${styles.content} ${showBody ? styles.contentExpanded : styles.contentCompact}${classNames?.content ? ` ${classNames.content}` : ""}`,
        style: isCenter ? { textAlign: "center" } : isRight ? { transform: "scaleX(-1)", textAlign: "right" } : { textAlign: "left" },
        children: [
          /* @__PURE__ */ jsx("div", { ref: headerRef, className: `${styles.header} ${titleColorMap[effectivePhase]}${classNames?.header ? ` ${classNames.header}` : ""}`, children: iconAndTitle }),
          /* @__PURE__ */ jsx(AnimatePresence, { children: showBody && hasDescription && !dismissing && /* @__PURE__ */ jsx(
            motion.div,
            {
              className: `${styles.description}${classNames?.description ? ` ${classNames.description}` : ""}`,
              style: { textAlign: "left" },
              initial: prefersReducedMotion ? false : { opacity: 0 },
              animate: { opacity: 1 },
              exit: { opacity: 0 },
              transition: prefersReducedMotion ? { duration: 0.01 } : { duration: 0.35, ease: [0.4, 0, 0.2, 1] },
              children: effectiveDescription
            },
            "description"
          ) }),
          /* @__PURE__ */ jsx(AnimatePresence, { children: showBody && hasAction && effectiveAction && !dismissing && /* @__PURE__ */ jsx(
            motion.div,
            {
              className: `${styles.actionWrapper}${classNames?.actionWrapper ? ` ${classNames.actionWrapper}` : ""}`,
              initial: prefersReducedMotion ? false : { opacity: 0 },
              animate: { opacity: 1 },
              exit: { opacity: 0 },
              transition: prefersReducedMotion ? { duration: 0.01 } : { duration: 0.35, ease: [0.4, 0, 0.2, 1], delay: 0.1 },
              children: /* @__PURE__ */ jsx(
                "button",
                {
                  className: `${styles.actionButton} ${actionColorMap[effectivePhase]}${classNames?.actionButton ? ` ${classNames.actionButton}` : ""}`,
                  onClick: handleActionClick,
                  type: "button",
                  "aria-label": effectiveAction.label,
                  children: effectiveAction.label
                }
              )
            },
            "action"
          ) })
        ]
      }
    )
  ] });
};
var ToastErrorBoundary = class extends Component {
  constructor() {
    super(...arguments);
    this.state = { hasError: false };
  }
  static getDerivedStateFromError() {
    return { hasError: true };
  }
  componentDidCatch(error, errorInfo) {
    if (process.env.NODE_ENV !== "production") {
      console.error("[GoeyToast] Rendering error:", error, errorInfo);
    }
  }
  render() {
    if (this.state.hasError) {
      return null;
    }
    return this.props.children;
  }
};
var DEFAULT_EXPANDED_DURATION = 4e3;
function GoeyToastWrapper({
  initialPhase,
  title,
  type,
  description,
  action,
  icon,
  classNames,
  fillColor,
  borderColor,
  borderWidth,
  timing,
  spring,
  bounce,
  toastId
}) {
  return /* @__PURE__ */ jsx(ToastErrorBoundary, { children: /* @__PURE__ */ jsx(
    GoeyToast,
    {
      title,
      description,
      type,
      action,
      icon,
      phase: initialPhase,
      classNames,
      fillColor,
      borderColor,
      borderWidth,
      timing,
      spring,
      bounce,
      toastId
    }
  ) });
}
function PromiseToastWrapper({
  promise,
  data,
  toastId
}) {
  const [phase, setPhase] = useState("loading");
  const [title, setTitle] = useState(data.loading);
  const [description, setDescription] = useState(data.description?.loading);
  const [action, setAction] = useState(void 0);
  useEffect(() => {
    const resetDuration = (hasExpandedContent) => {
      const baseDuration = data.timing?.displayDuration ?? (hasExpandedContent ? DEFAULT_EXPANDED_DURATION : void 0);
      const collapseDurMs = 0.9 * 1e3;
      const duration = baseDuration != null && hasExpandedContent ? baseDuration + collapseDurMs : baseDuration;
      if (duration != null) {
        toast.custom(() => /* @__PURE__ */ jsx(PromiseToastWrapper, { promise, data, toastId }), { id: toastId, duration });
      }
    };
    promise.then((result) => {
      const desc = typeof data.description?.success === "function" ? data.description.success(result) : data.description?.success;
      setTitle(
        typeof data.success === "function" ? data.success(result) : data.success
      );
      setDescription(desc);
      setAction(data.action?.success);
      setPhase("success");
      resetDuration(Boolean(desc || data.action?.success));
    }).catch((err) => {
      const desc = typeof data.description?.error === "function" ? data.description.error(err) : data.description?.error;
      setTitle(
        typeof data.error === "function" ? data.error(err) : data.error
      );
      setDescription(desc);
      setAction(data.action?.error);
      setPhase("error");
      resetDuration(Boolean(desc || data.action?.error));
    });
  }, []);
  return /* @__PURE__ */ jsx(ToastErrorBoundary, { children: /* @__PURE__ */ jsx(
    GoeyToast,
    {
      title,
      description,
      type: phase === "loading" ? "info" : phase,
      action,
      phase,
      classNames: data.classNames,
      fillColor: data.fillColor,
      borderColor: data.borderColor,
      borderWidth: data.borderWidth,
      timing: data.timing,
      spring: data.spring,
      bounce: data.bounce
    }
  ) });
}
function createGoeyToast(title, type, options) {
  const hasExpandedContent = Boolean(options?.description || options?.action);
  const baseDuration = options?.timing?.displayDuration ?? options?.duration ?? (options?.description ? DEFAULT_EXPANDED_DURATION : void 0);
  const duration = hasExpandedContent ? Infinity : baseDuration;
  const toastId = options?.id ?? Math.random().toString(36).slice(2);
  return toast.custom(
    () => /* @__PURE__ */ jsx(
      GoeyToastWrapper,
      {
        initialPhase: type,
        title,
        type,
        description: options?.description,
        action: options?.action,
        icon: options?.icon,
        classNames: options?.classNames,
        fillColor: options?.fillColor,
        borderColor: options?.borderColor,
        borderWidth: options?.borderWidth,
        timing: options?.timing,
        spring: options?.spring,
        bounce: options?.bounce,
        toastId: hasExpandedContent ? toastId : void 0
      }
    ),
    {
      duration,
      id: toastId
    }
  );
}
var goeyToast = Object.assign(
  (title, options) => createGoeyToast(title, "default", options),
  {
    success: (title, options) => createGoeyToast(title, "success", options),
    error: (title, options) => createGoeyToast(title, "error", options),
    warning: (title, options) => createGoeyToast(title, "warning", options),
    info: (title, options) => createGoeyToast(title, "info", options),
    promise: (promise, data) => {
      const id = Math.random().toString(36).slice(2);
      return toast.custom(() => /* @__PURE__ */ jsx(PromiseToastWrapper, { promise, data, toastId: id }), {
        id,
        duration: data.timing?.displayDuration != null || data.description ? Infinity : void 0
      });
    },
    dismiss: toast.dismiss
  }
);

export { GoeyToaster, goeyToast };
